File: src/main/java/graph/edge/DuplexChannel.java
package graph.edge;

import graph.vertex.Vertex;

public class DuplexChannel extends AbstractEdge {
    private DuplexChannel pair;

    private DuplexChannel(DuplexChannel channel) {
        super(channel.toNode, channel.fromNode, channel.weight, channel.errors);
        pair = channel;
    }

    public DuplexChannel(Vertex fromNode, Vertex toNode, int weight, double errors) {
        super(fromNode, toNode, weight, errors);
        pair = new DuplexChannel(this);
        connect();
    }

    @Override
    public void connect() {
        super.connect();
        toNode.addConnection(pair);
    }

    @Override
    public void remove() {
        super.remove();
        toNode.removeConnection(pair);
    }

    @Override
    protected String getDirection() {
        return "<==>";
    }
}

File: src/main/java/mvc/model/channel/DuplexChannel.kt
package mvc.model.channel

import mvc.model.node.Node

class DuplexChannel : AbstractChannel {
    private var pair: DuplexChannel

    override val direction = "<==>"

    private constructor(channel: DuplexChannel) : super(channel.toNode, channel.fromNode, channel.weight, channel.errors) {
        pair = channel
    }

    constructor(fromNode: Node, toNode: Node, weight: Int, errors: Double) : super(fromNode, toNode, weight, errors) {
        pair = DuplexChannel(this)
        connect()
    }

    override fun connect() {
        super.connect()
        toNode.addConnection(pair)
    }

    override fun remove() {
        super.remove()
        toNode.removeConnection(pair)
    }
}
